docs/blueprints/slow-down-and-retry.md

```diff
@@ -0,0 +1,84 @@
+
+Upload operation combines several requests:
+```
+  begin object upload
+    (for remote segment)
+      begin segment upload 
+        send data to storage node
+      commit segment upload
+    (or for inline segment)
+      make inline segment
+    ... repeat for multiple segments
+  commit object upload
+```
+
+Requests are batched:
+1. begin object upload + (first begin segment upload or make inline segment)
+2. commit segment upload + (begin segment upload or make inline segment) for multi-segment object
+3. (last commit segment upload or last make inline segment) + commit object upload
+4. special case for multi-segment object with last segment inline: commit segment upload + last make inline segment + commit object upload
+
+In case of `Too Many Requests` during upload operation we should be able to retry batched requests as soon as we resolve point 1 from Open Issues. Other case for upload is when we fail to upload single segment to storage nodes then we should retry uploading whole segment from scratch. In future we can improve this logic to re-upload failed pieces to different storage nodes.
+
+In case of `Too Many Requests` during download operation we should be able to retry single Metainfo API requests. If downloading segment from storage nodes will fail we should retry downloading this single segment.
+
+As an addition to retry logic we need to implement at least basic backoff mechanism that will increase the waiting time (between retries) up to a certain threshold. This will prevent from overwhelming satellite with retry requests. For our needs exponential backoff looks to be reasonable solution. 
+
+Additional rules:
+* Every retried request should be processed by backoff logic.
+* Retry and slow mechanism should be transient from libuplink consumer perspective except several necessary configurable options (see Implementation).
+* Applying retry and slow down logic shouldn't degrade libuplink overall performance and increase number of round trips to satellite.
+
+## Implementation
+
+TODO: will be added when general design will be accepted.
+
+Options:
+* initial interval
+* max elapsed time or number of retries ??
+* multipliter
+* randomization factor
+* max interval
+
+## Open issues
+
+1. Currently, endpoints handled by DRPC in case of error are not returning value, only error. With Metainfo `Batch` call we would need to have partial results which requests were executed successfully. 
+2. Can we reuse order limits from failed upload? Failed upload means when sending data to storage nodes was unsuccessful. 
```

Comment:
I am not sure what this means. The idea to have a retry logic is to avoid failed uploads altogether, isn't it?

Reply:
Yes but except issues with Metainfo API we can also fail during upload (e.g. large number of selected nodes can be broken/unavailable). If last from 100 uploaded segments will fail it would be more convenient to reupload it instead of failing whole upload. At the moment I'm just not sure if we are able to reuse order limits that were used for such failed segment upload or we need to ask satellite for a new one.